### ISSUE TITLE
libfl.so: undefined reference to `yylex'

### ISSUE THREAD

**Issue by eeach** [Seifert69/DikuMUD3#70]
Is anyone else having trouble compiling with this error? I did some googling and at first thought the problem was with flex, but it hasn't updated in 2+ years. I don't have much experience with C++, so I'm probably missing something simple. :-|

I'm doing this in Docker, so it's all fresh. I started with Arch Linux, but moved to Debian when I first ran into problems.  Any help is greatly appreciated. Thanks! :-)

```
root@f7e15a47dfc0:/diku/vme/src# uname -a; apt list --installed | grep 'gcc\|g++\|flex\|bison\|libboost-all-dev' 
Linux f7e15a47dfc0 5.8.12-arch1-1 #1 SMP PREEMPT Sat, 26 Sep 2020 21:42:58 +0000 x86_64 GNU/Linux

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

bison/stable,now 2:3.3.2.dfsg-1 amd64 [installed]
flex/stable,now 2.6.4-6.2 amd64 [installed]
g++-8/stable,now 8.3.0-6 amd64 [installed,automatic]
g++/stable,now 4:8.3.0-1 amd64 [installed]
gcc-8-base/stable,now 8.3.0-6 amd64 [installed,automatic]
gcc-8/stable,now 8.3.0-6 amd64 [installed,automatic]
gcc/stable,now 4:8.3.0-1 amd64 [installed,automatic]
libbison-dev/stable,now 2:3.3.2.dfsg-1 amd64 [installed,automatic]
libboost-all-dev/stable,now 1.67.0.1 amd64 [installed]
libgcc-8-dev/stable,now 8.3.0-6 amd64 [installed,automatic]
libgcc1/stable,now 1:8.3.0-6 amd64 [installed,automatic]
```

```
root@f7e15a47dfc0:/diku/vme/src# make all
g++ -pipe -m64 -o ../bin/vme  ../build/./comm.o ../build/./config.o ../build/./apf_affect.o ../build/./utility.o ../build/./namelist.o ../build/./queue.o ../build/./dbfind.o ../build/./experience.o ../build/./act_color.o ../build/./hookmud.o ../build/./fight.o ../build/./structs.o ../build/./act_change.o ../build/./nanny.o ../build/./convert.o ../build/./main.o ../build/./dictionary.o ../build/./tif_affect.o ../build/./slime.o ../build/./db_file.o ../build/./cmdload.o ../build/./bytestring.o ../build/./interpreter.o ../build/./bank.o ../build/./dilfld.o ../build/./color.o ../build/./teach.o ../build/./spell_parser.o ../build/./act_info.o ../build/./trie.o ../build/./reception.o ../build/./event.o ../build/./dilrun.o ../build/./nice.o ../build/./dilshare.o ../build/./destruct.o ../build/./mobact.o ../build/./magic.o ../build/./signals.o ../build/./act_other.o ../build/./act_wstat.o ../build/./act_skills.o ../build/./combat.o ../build/./badnames.o ../build/./zon_wiz.o ../build/./pcsave.o ../build/./money.o ../build/./spec_assign.o ../build/./system.o ../build/./sector.o ../build/./common.o ../build/./unitfind.o ../build/./hook.o ../build/./modify.o ../build/./dilinst.o ../build/./dilexp.o ../build/./skills.o ../build/./justice.o ../build/./eliza.o ../build/./intlist.o ../build/./zone_reset.o ../build/./db.o ../build/./act_wizard.o ../build/./weather.o ../build/./spec_procs.o ../build/./affect.o ../build/./handler.o ../build/./guild.o ../build/./zon_basis.o ../build/./act_movement.o ../build/./dilsup.o ../build/./extra.o ../build/./account.o ../build/./ban.o ../build/./textutil.o ../build/./limits.o ../build/./act_offensive.o ../build/./protocol.o ../build/./db_utils.o ../build/./files.o ../build/./membug.o ../build/./constants.o ../build/./path.o ../build/./str_parse.o -lcrypt -lpthread -lboost_filesystem  -lboost_regex -lboost_system  
g++ -pipe -m64 -o ../bin/mplex  ../build/./mplex2/network.o ../build/./mplex2/translate.o ../build/./mplex2/ClientConnector.o ../build/./mplex2/mplex.o ../build/./mplex2/echo_server.o ../build/./mplex2/MUDConnector.o ../build/hook.o ../build/queue.o ../build/textutil.o ../build/utility.o ../build/common.o ../build/protocol.o ../build/color.o ../build/bytestring.o ../build/membug.o -lcrypt -lpthread -lboost_filesystem  -lboost_regex -lboost_system  
g++ -pipe -m64 -o ../bin/defcomp  ../build/./defcomp/defcomp.o -lcrypt -lpthread -lboost_filesystem  -lboost_regex -lboost_system  
g++ -pipe -m64 -o ../bin/pp  ../build/./pp/pp6.o ../build/./pp/pp2.o ../build/./pp/pp3.o ../build/./pp/pp4.o ../build/./pp/pp8.o ../build/./pp/pp5.o ../build/./pp/main.o ../build/./pp/pp7.o ../build/./pp/pp1.o -lcrypt -lpthread -lboost_filesystem  -lboost_regex -lboost_system  
cd vmc;make -j8
make[1]: Entering directory '/diku/vme/src/vmc'
g++  -I.  -I.. -DVMC_SRC -DVMC  -DLINUX -DPOSIX -pipe -m64 -DMUD_DEBUG -g -Wall  -g -c vmc.cpp -o ../../build//vmc/vmc.o
g++  -I.  -I.. -DVMC_SRC -DVMC  -DLINUX -DPOSIX -pipe -m64 -DMUD_DEBUG -g -Wall  -g -c vmc_process.cpp -o ../../build//vmc/vmc_process.o
g++  -I.  -I.. -DVMC_SRC -DVMC  -DLINUX -DPOSIX -pipe -m64 -DMUD_DEBUG -g -Wall  -g -c utility.cpp -o ../../build//vmc/utility.o
g++  -I.  -I.. -DVMC_SRC -DVMC  -DLINUX -DPOSIX -pipe -m64 -DMUD_DEBUG -g -Wall  -g -c structs.cpp -o ../../build//vmc/structs.o
g++  -I.  -I.. -DVMC_SRC -DVMC  -DLINUX -DPOSIX -pipe -m64 -DMUD_DEBUG -g -Wall  -g -c destruct.cpp -o ../../build//vmc/destruct.o
g++  -I.  -I.. -DVMC_SRC -DVMC  -DLINUX -DPOSIX -pipe -m64 -DMUD_DEBUG -g -Wall  -g -c dilshare.cpp -o ../../build//vmc/dilshare.o
g++  -I.  -I.. -DVMC_SRC -DVMC  -DLINUX -DPOSIX -pipe -m64 -DMUD_DEBUG -g -Wall  -g -c membug.cpp -o ../../build//vmc/membug.o
g++  -I.  -I.. -DVMC_SRC -DVMC  -DLINUX -DPOSIX -pipe -m64 -DMUD_DEBUG -g -Wall  -g -c spec_assign.cpp -o ../../build//vmc/spec_assign.o
g++  -I.  -I.. -DVMC_SRC -DVMC  -DLINUX -DPOSIX -pipe -m64 -DMUD_DEBUG -g -Wall  -g -c money.cpp -o ../../build//vmc/money.o
g++  -I.  -I.. -DVMC_SRC -DVMC  -DLINUX -DPOSIX -pipe -m64 -DMUD_DEBUG -g -Wall  -g -c pp1.cpp -o ../../build//vmc/pp1.o
g++  -I.  -I.. -DVMC_SRC -DVMC  -DLINUX -DPOSIX -pipe -m64 -DMUD_DEBUG -g -Wall  -g -c pp2.cpp -o ../../build//vmc/pp2.o
g++  -I.  -I.. -DVMC_SRC -DVMC  -DLINUX -DPOSIX -pipe -m64 -DMUD_DEBUG -g -Wall  -g -c pp3.cpp -o ../../build//vmc/pp3.o
g++  -I.  -I.. -DVMC_SRC -DVMC  -DLINUX -DPOSIX -pipe -m64 -DMUD_DEBUG -g -Wall  -g -c pp4.cpp -o ../../build//vmc/pp4.o
g++  -I.  -I.. -DVMC_SRC -DVMC  -DLINUX -DPOSIX -pipe -m64 -DMUD_DEBUG -g -Wall  -g -c pp5.cpp -o ../../build//vmc/pp5.o
g++  -I.  -I.. -DVMC_SRC -DVMC  -DLINUX -DPOSIX -pipe -m64 -DMUD_DEBUG -g -Wall  -g -c pp6.cpp -o ../../build//vmc/pp6.o
g++  -I.  -I.. -DVMC_SRC -DVMC  -DLINUX -DPOSIX -pipe -m64 -DMUD_DEBUG -g -Wall  -g -c pp7.cpp -o ../../build//vmc/pp7.o
g++  -I.  -I.. -DVMC_SRC -DVMC  -DLINUX -DPOSIX -pipe -m64 -DMUD_DEBUG -g -Wall  -g -c pp8.cpp -o ../../build//vmc/pp8.o
g++  -I.  -I.. -DVMC_SRC -DVMC  -DLINUX -DPOSIX -pipe -m64 -DMUD_DEBUG -g -Wall  -g -c vmc_db_file.cpp -o ../../build//vmc/vmc_db_file.o
bison --debug -y -d -v vmcpar.y
vmcpar.y: warning: 2 shift/reduce conflicts [-Wconflicts-sr]
flex -t vmclex.l > ../../build//vmc/tmp_vmclex.cpp
bison --debug -d -v -pdil dilpar.y
flex -Pdil -t dillex.l  > ../../build//vmc/tmp_dillex.cpp 
g++  -I.  -I.. -DVMC_SRC -DVMC  -DLINUX -DPOSIX -pipe -m64 -DMUD_DEBUG -g -Wall  -g -I ../../build//vmc/ -c ../../build//vmc/tmp_vmcpar.cpp -o ../../build//vmc/tmp_vmcpar.o
g++  -I.  -I.. -DVMC_SRC -DVMC  -DLINUX -DPOSIX -pipe -m64 -DMUD_DEBUG -g -Wall  -g -I ../../build//vmc/ -c ../../build//vmc/tmp_vmclex.cpp -o ../../build//vmc/tmp_vmclex.o
g++  -I.  -I.. -DVMC_SRC -DVMC  -DLINUX -DPOSIX -pipe -m64 -DMUD_DEBUG -g -Wall  -g -I ../../build//vmc/ -c ../../build//vmc/tmp_dilpar.cpp -o ../../build//vmc/tmp_dilpar.o
g++  -I.  -I.. -DVMC_SRC -DVMC  -DLINUX -DPOSIX -pipe -m64 -DMUD_DEBUG -g -Wall  -g -I ../../build//vmc/ -c ../../build//vmc/tmp_dillex.cpp -o ../../build//vmc/tmp_dillex.o
g++ -I. -I.. -I vmc -DVMC  -DLINUX -DPOSIX -pipe -m64 -DMUD_DEBUG -g -Wall   -O0 ../../build//vmc/vmc.o ../../build//vmc/vmc_process.o ../../build//vmc/utility.o ../../build//vmc/structs.o ../../build//vmc/destruct.o ../../build//vmc/dilshare.o ../../build//vmc/membug.o ../../build//vmc/spec_assign.o ../../build//vmc/money.o ../../build//vmc/pp1.o ../../build//vmc/pp2.o ../../build//vmc/pp3.o ../../build//vmc/pp4.o ../../build//vmc/pp5.o ../../build//vmc/pp6.o ../../build//vmc/pp7.o ../../build//vmc/pp8.o ../../build//vmc/vmc_db_file.o ../../build//vmc/tmp_vmcpar.o ../../build//vmc/tmp_vmclex.o ../../build//vmc/tmp_dilpar.o ../../build//vmc/tmp_dillex.o ../../build//color.o ../../build//textutil.o ../../build//namelist.o ../../build//files.o ../../build//extra.o ../../build//bytestring.o ../../build//common.o ../../build//intlist.o -lfl -g -o ../../bin/vmc 
/usr/bin/ld: /usr/lib/gcc/x86_64-linux-gnu/8/../../../x86_64-linux-gnu/libfl.so: undefined reference to `yylex'
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:92: ../../bin/vmc] Error 1
make[1]: Leaving directory '/diku/vme/src/vmc'
make: *** [Makefile:211: ../bin/vmc] Error 2
```

**Comment by Seifert69** [comment]
Odd. I've never seen anything like it

Try to look here 

https://stackoverflow.com/questions/1480138/undefined-reference-to-yylex

Let me know if any of those tips work plz and we can update the Makefile

**Comment by Seifert69** [comment]
I tried to add in the extern statement to test your issue. It still compiles fine on mine.

Try to pull the latest
** be sure to "make clean" **

Then

make all -j8

LMK how it goes plz :)

**Comment by eeach** [comment]
Working this now. Give me a bit. Thanks for the quick response! :-)

**Comment by eeach** [comment]
I played with all the solutions on that SO page. The only one that I got to work was installing flex-old ( 2.5.4a-10 ). 

Ran into some trouble with zone files, just moved the ones that it complained about... Got diku to start, complain about my telnet connection to the websocket and complain about missing zones. So, it looks like it's working. :-)

**Comment by wilminator** [comment]
@eeach, what image were you building your MUD in?  I have mine working in an Ubuntu image. There I only had to install flex.

**Comment by wilminator** [comment]
@eeach nm, just caught that you were using Debian.

**Comment by Seifert69** [comment]
@eeach we'd love to know what combo caused the issues :-) just to add to
docs.

**Comment by Seifert69** [comment]
This is still an issue on Debian and probably other systems. The solution for now is to use flex-old. However, I'd really like a more permanent solution :-) Any help appreciated

**Comment by serlin** [comment]
Hi there, 

I ran into this issue on Arch, which afaict does not keep a flex-old package so I dug in a bit deeper. I wound up trying a couple of solutions linked above or otherwise google-able. There were two insights worth sharing:

In the generated `.../vmc/link.txt` doing either `-Wl,--as-needed,/usr/lib/libfl.so` or removing the explicit reference to `/usr/lib/libfl.so` allowed it to compile and it ran fine. I tried chasing the stack up and the best I could come up with was to, somewhat puzzlingly, remove: `${FLEX_LIBRARIES}` from CMakeLists.txt in `src/vmc`. I don't know if there are side effects to that that or if that would still work for those in Ubuntu.
